// modex/controllers/BasicCrud.js
"use strict";
/**
 * Default actions:
 *  - GET : list  show  new  edit  confirmDelete
 *  - POST: create, modify
 *  - PUT : update, multiUpdate
 *  - DELETE : delete  multiDelete
 * Options:
 *   name : String : Name of the controller (Optional)
 *   model : Object - A moongose model or any other types of model that provide compatible interface. (Required)
 *   actions : String - Specify default handlers. Use '-' prefix to specify exclusion list.
 *     Eg: 'list show' will only include list and show. '- new delete' will exclude both new and delete. Default will include all default handlers. (Optional)
 *   defaultFormat: String - A choice of the following: 'json', 'html', 'xml'. Default is 'html'. (Optional)
 *   allowedFormats : [String] - Allowed view types that can be rendered. Eg: ['html']. Default is none. (Optional)
 */
 
var formatFields = require('../helper.model').formatFields;
 

function init() {
  // Setup displayed fields for each controller action
  var map = this._fieldArr;
  for (var key in this) {
    if (typeof this[key] === 'function') {
      var fieldStr = this[key+'_fields'];
      if (!fieldStr) fieldStr = '';
      if (!(fieldStr in map)) map[fieldStr] = formatFields(this._model.schema, fieldStr);
    }
  }
}

function attachAsArrayFunction(controller, actionName) {
  controller._model.prototype.asArray = function() {
    var data = [], doc = this;
    var fieldStr = controller[actionName+'_fields'];
    controller._fieldArr[fieldStr].forEach(function(f) {
      data.push(doc[f]);
    });
    return data;
  };
}

function populateSelection(action) {

}


//TODO: create function to generate url from relative path
function resolveRedirect(str) {
  if (str === 'back') return str;
  
}

/** Handles mongoose errors */
function handleError(err, next) {
  //console.log(err);
  if (err.name === 'CastError' && err.type === 'ObjectId') next();
  else return next(err);
}

function createOptions(controller, actionName, records, options) {
  options = options || {};
  options._controller = controller;
  options._actionName = actionName;
  options._modelName = controller._model.modelName, 
  options._view = controller[actionName+'_view'] || actionName;
  options._layout = controller[actionName+'_layout'] || controller._layout || ['controllerLayout', 'layout'];
  options.data = {records:records, fields:controller._fieldArr[controller[actionName+'_fields']]};
  options.title = actionName + ' ' + options._modelName;
  return options;
}

var defaultActions = ['list', 'new', 'create', 'edit', 'confirmDelete', 'show', 'update', 'multiUpdate', 'delete', 'multiDelete', 'modify'];

var BasicCrud = module.exports = function(params) {
  require('./Base').call(this, params);
  var actions = [];
  if (params.actions) {
    var sActions = params.actions.split(/\s+/);
    if (sActions[0] === '-') {
      defaultActions.forEach(function(a) {
        if (sActions.indexOf(a) >= 0) actions.push(a);
      });
    } else {
      sActions.forEach(function(a) {
        //if (defaultActions.indexOf(a) >= 0) 
        actions.push(a);
      });
    }
  } else {
    actions = defaultActions;
  }
  Object.defineProperties(this, {
    _model : {value:params.model},
    _fieldArr : {value:{}},
    _actions : {value: actions},
  });
};

Object.defineProperties(BasicCrud.prototype, {
  _init : {value: init}
});


BasicCrud.prototype.list = function(req, res, next){
  var p = req.query.p || 1; // page number
  var pp = req.query.pp || 100; // number of record per page
  var sort = req.query.sort; // sort parameter, eg: sort='name -email'
  var filter;
  if (typeof this.list_filter === 'function') filter = this.list_filter(req);
  else filter = this.list_filter;
  var controller = this;
  this._model.find(filter, this.list_fields, {sort: sort, skip:(p-1)*pp, limit:pp}, function(err, records) {
    if (err) return handleError(err, next);
    attachAsArrayFunction(controller, 'list');
    req.format = controller.list_format || req.format;
    var options = createOptions(controller, 'list', records);
      //actionMenu: [{label:'New', link:controller.controllerName+controller.new_route}],
      //_route: req.url
    if (typeof controller.list_process === 'function') controller.list_process(options);
    res.respond(options);
    //menus: activate, deactivate, delete
  });
};
BasicCrud.prototype.list_method = 'get';
BasicCrud.prototype.list_route = '';
BasicCrud.prototype.list_fields = '';


BasicCrud.prototype.new = function(req, res, next){
  attachAsArrayFunction(this, 'new');
  var options = createOptions(this, 'new', null, {
    _controllerPath:req.route.path.slice(0, -4)
  });
  res.respond(options);
};
BasicCrud.prototype.new_method = 'get';
BasicCrud.prototype.new_route = '.new';
BasicCrud.prototype.new_fields = '';


BasicCrud.prototype.create = function(req, res, next){
  var controller = this;
  var method = req.param('_method', 'post');//default to create
  if (method === 'put') {
    this.multiUpdate(req, res, next);
  } else if (method === 'delete') {
    req.forwardedByController = true;
    this.multiDelete(req, res, next);
  } else {
    this._model.create(req.body, function(err) {
      if (err) return handleError(err, next);
      //res.message('Information updated!');
      var redirpath = req.body._redirect;
      var options = createOptions(controller, 'create', null, {_redirect:redirpath});
      res.respond(options);
    });
  }
};
BasicCrud.prototype.create_method = 'post';
BasicCrud.prototype.create_route = '';


function commonEditOrShow(req, res, next, controller, actionName) {
  controller._model.findById(req.params.id, function(err, doc) {
    if (err) return handleError(err, next);
    attachAsArrayFunction(controller, actionName);
    var options = createOptions(controller, actionName, [doc]);
    res.respond(options);
  });
}

// Html form will do post to /module/controller/:id
BasicCrud.prototype.edit = function(req, res, next){
  commonEditOrShow(req, res, next, this, 'edit');
};
BasicCrud.prototype.edit_method = 'get';
BasicCrud.prototype.edit_route = '/:id.edit'; 
BasicCrud.prototype.edit_fields = '';


BasicCrud.prototype.confirmDelete = function(req, res, next){
  commonEditOrShow(req, res, next, this, 'confirmDelete');
};
BasicCrud.prototype.confirmDelete_method = 'get';
BasicCrud.prototype.confirmDelete_route = '/:id.delete';
BasicCrud.prototype.confirmDelete_fields = '';


// Show or display edit form
BasicCrud.prototype.show = function(req, res, next){
  commonEditOrShow(req, res, next, this, 'show');
};
BasicCrud.prototype.show_method = 'get';
BasicCrud.prototype.show_route = '/:id';
BasicCrud.prototype.show_fields = '';


BasicCrud.prototype.update = function(req, res, next){
  var controller = this;
  //console.log(req.body);
  var options = createOptions(controller, 'update', null, {_redirect:req.body._redirect});
  res.respond(options);
  this._model.findByIdAndUpdate(req.params.id, { $set: req.body }, function(err, count) {
    if (err) return handleError(err, next);
    //res.message('Information updated!');
    //res.redirect('back');
    var options = createOptions(controller, 'update', null, {count:count, _redirect:req.body._redirect});
    res.respond(options);
  });
};
BasicCrud.prototype.update_method = 'put';
BasicCrud.prototype.update_route = '/:id';


BasicCrud.prototype.multiUpdate = function(req, res, next){
  var controller = this;
  var conditions = {};
  if (req.body._selection && req.body._selection.length > 0) {
    conditions._id = {$in:req.body._selection};
    delete req.body._selection;
  }
  this._model.update(conditions, {$set:req.body}, {multi: true}, function(err, count) {
    if (err) return handleError(err, next);
    var options = createOptions(controller, 'multiUpdate', null, {count:count, _redirect:req.body._redirect});
    res.respond(options);
  });
};
BasicCrud.prototype.multiUpdate_method = 'put';
BasicCrud.prototype.multiUpdate_route = '';
BasicCrud.prototype.multiUpdate_view = 'update';

/// Delete one record
BasicCrud.prototype.delete = function(req, res, next){
  var controller = this;
  this._model.remove({_id:req.params.id}, function(err) {
    if (err) return handleError(err, next);
    if (req.forwardedByController) res.respond(createOptions(controller, 'delete', null, {_redirect:req.body._redirect}));
    else res.send(req.body);
  });
};
BasicCrud.prototype.delete_method = 'delete';
BasicCrud.prototype.delete_route = '/:id';

/// Delete all/multiple records
BasicCrud.prototype.multiDelete = function(req, res, next){
  var controller = this;
  var conditions = {};
  if (req.body._selection && req.body._selection.length > 0) {
    conditions._id = {$in:req.body._selection};
    delete req.body._selection;
  }
  this._model.remove(conditions, function(err) {
    if (err) return handleError(err, next);
    if (req.forwardedByController) res.respond(createOptions(controller, 'multiDelete', null, {_redirect:req.body._redirect}));
    else res.send(req.body);
  });
};
BasicCrud.prototype.multiDelete_method = 'delete';
BasicCrud.prototype.multiDelete_route = '';
BasicCrud.prototype.multiUpdate_view = 'delete';

//Update and Delete proxy
BasicCrud.prototype.modify = function(req, res, next){
  var method = req.param('_method', 'put');//default to update
  if (method === 'put') {
    this.update(req, res, next);
  } else if (method === 'delete') {
    req.forwardedByController = true;
    this.delete(req, res, next);
  } else {
    res.send(400, 'Unsupported Operation');
  }
};
BasicCrud.prototype.modify_method = 'post';
BasicCrud.prototype.modify_route = '/:id';

//BasicCrud.prototype.softDelete put :id

// add routes to default values
var defaults = require('./defaults');
for (var key in BasicCrud.prototype) {
  defaults.routes[key] = BasicCrud.prototype[key+'_route'];
}
